src/prefect/engine.py

```diff
@@ -2966,6 +2966,15 @@ async def create_autonomous_task_run(task: Task, parameters: Dict[str, Any]) ->
             factory = await ResultFactory.from_autonomous_task(task, client=client)
             await factory.store_parameters(parameters_id, parameters)
 
+        if not task.task_origin_hash:
+            raise ValueError(
+                f"The source file of task {task.name!r} is not available. This is"
+                " required for autonomous task runs in order to determine which"
+                " `TaskServer` instance is capable of running the task."
+            )
+
+        task.task_key = task.task_origin_hash
```

Comment:
Hmm, won't this obscure the task key for _all_ functions?  I thought we only wanted to do this for ones that fell into the `__main__` trap?

Reply:
it would not modify the task key on `Task` object itself, but we would use this new hash as the task key on every autonomous `TaskRun` - I was thinking it felt strange to do modify the task key on the object for some subset of tasks and not for another. It didn't feel like a big concession that the task keys on the task runs were not human readable - what do you think?

if you prefer what we had mentioned earlier I can switch it back

Reply:
hmm yeah it does, updated in [7507ba3](https://github.com/PrefectHQ/prefect/pull/12227/commits/7507ba38cd1280b2237b1ffca2246257f384d905)